src/libraries/System.Linq/tests/ShuffleTests.cs

```diff
@@ -0,0 +1,316 @@
+                Array.Sort(shuffled);
+                Assert.Equal(Enumerable.Range(0, length), shuffled);
+
+                shuffled = source.Shuffle().Take(length).ToArray();
+                Assert.Equal(length, shuffled.Length);
+                Array.Sort(shuffled);
+                Assert.Equal(Enumerable.Range(0, length), shuffled);
+            });
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(1)]
+        [InlineData(2)]
+        [InlineData(3)]
+        [InlineData(100)]
+        public void ToList_AllElementsReturned(int length)
+        {
+            Assert.All(CreateSources(Enumerable.Range(0, length)), source =>
+            {
+                List<int> shuffled;
+
+                shuffled = source.Shuffle().ToList();
+                Assert.Equal(length, shuffled.Count);
+                shuffled.Sort();
+                Assert.Equal(Enumerable.Range(0, length), shuffled);
+
+                shuffled = source.Shuffle().Take(length).ToList();
+                Assert.Equal(length, shuffled.Count);
+                shuffled.Sort();
+                Assert.Equal(Enumerable.Range(0, length), shuffled);
+            });
+        }
+
+        [Fact]
+        public void Enumeration_ElementsAreRandomized()
+        {
+            // The chance that shuffling a thousand elements produces the same order twice is infinitesimal.
+            int length = 1000;
+            Assert.All(CreateSources(Enumerable.Range(0, length)), source =>
+            {
+                List<int> first, second;
+
+                first = [];
+                second = [];
+                foreach (int i in source.Shuffle()) first.Add(i);
+                foreach (int i in source.Shuffle()) second.Add(i);
+                Assert.Equal(length, first.Count);
+                Assert.Equal(length, second.Count);
+                Assert.NotEqual(first, second);
+
+                foreach (int takeCount in new[] { length - 1, length + 1 })
+                {
+                    first = [];
+                    second = [];
+                    foreach (int i in source.Shuffle().Take(takeCount)) first.Add(i);
+                    foreach (int i in source.Shuffle().Take(takeCount)) second.Add(i);
+                    Assert.Equal(Math.Min(takeCount, length), first.Count);
+                    Assert.Equal(Math.Min(takeCount, length), second.Count);
+                    Assert.NotEqual(first, second);
+
+                    first = [];
+                    second = [];
+                    foreach (int i in source.Shuffle().Take(takeCount + 1).Take(takeCount)) first.Add(i);
+                    foreach (int i in source.Shuffle().Take(takeCount + 1).Take(takeCount)) second.Add(i);
+                    Assert.Equal(Math.Min(takeCount, length), first.Count);
+                    Assert.Equal(Math.Min(takeCount, length), second.Count);
+                    Assert.NotEqual(first, second);
```

Comment:
Is it worth asserting that first.Distinct().Count() is also equal to the requested count?  That is, that none of "too short", "too long", (or, elsewhere "just right") has a path that duplicates entries?